scipio/src/timer.rs

```diff
@@ -133,6 +140,179 @@ impl Future for Timer {
+        when: Instant,
+        action: impl Future<Output = T> + 'static,
+    ) -> TimerAction<T> {
+        let task = Task::local(async move {
+            let now = Instant::now();
+            if when > now {
+                let dur = when.duration_since(now);
+                Timer::new(dur).await;
+            }
+            action.await
+        });
+
+        TimerAction {
+            handle: task.detach(),
+        }
+    }
+
+    /// Creates a [`TimerAction`] that will execute the associated future once after some
+    /// time is passed
+    ///
+    /// # Arguments
+    ///
+    /// * `period` a [`Duration`] that represents the period in which the action repeats
+    /// * `action` a Future to be executed after `when` is elapsed.
+    ///
+    /// # Examples
+    ///
+    /// ```no_run
+    /// use scipio::{LocalExecutor, TimerAction};
+    /// use std::time::Duration;
+    ///
+    /// LocalExecutor::spawn_executor("test", None, || async move {
+    ///     let action = TimerAction::periodic(Duration::from_millis(100), || async move {
+    ///         println!("Execute this forever");
+    ///     });
+    ///     action.join().await; // this never returns
+    /// });
+    /// ```
+    /// [`Duration`]: https://doc.rust-lang.org/std/time/struct.Duration.html
+    /// [`TimerAction`]: struct.TimerAction
+    pub fn periodic<G, F>(period: Duration, action_gen: G) -> TimerAction<()>
+    where
+        G: Fn() -> F + 'static,
+        F: Future<Output = T> + 'static,
+    {
+        let task = Task::local(async move {
+            loop {
+                Timer::new(period).await;
+                action_gen().await;
+            }
+        });
+
+        TimerAction {
+            handle: task.detach(),
+        }
+    }
+
+    /// Cancel an existing [`TimerAction`]
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use scipio::{LocalExecutor, TimerAction};
+    /// use std::time::Duration;
+    ///
+    /// LocalExecutor::spawn_executor("test", None, || async move {
+    ///     let action = TimerAction::periodic(Duration::from_millis(100), || async move {
+    ///         println!("Execute this forever");
+    ///     });
+    ///     action.cancel().await;
+    /// });
+    /// ```
+    /// [`TimerAction`]: struct.TimerAction
+    pub async fn cancel(self) {
+        self.handle.cancel();
+        self.handle.await;
+    }
+
+    /// Waits for a [`TimerAction`] to return
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use scipio::{LocalExecutor, TimerAction};
+    /// use std::time::Duration;
+    ///
+    /// LocalExecutor::spawn_executor("test", None, || async move {
+    ///     let action = TimerAction::once(Duration::from_millis(100), async move {
+    ///         println!("Execute this in 100ms");
+    ///     });
+    ///     action.join().await;
+    /// });
+    /// ```
+    /// [`TimerAction`]: struct.TimerAction
+    pub async fn join(self) {
+        self.handle.await;
+    }
```

Comment:
How does that work with repeated timers? Does it returns it the timer is cancelled in another fiber?

Reply:
this should not be used with repeated timers. The problem is that we can't (due to way rust works) both `cancel` and `join`: `cancel` takes a `self` parameter, not a reference, so after you call it you cannot do anything with the `JoinHandle` anymore.

We could change it like we changed `detach` but that is a bit more complex because `cancel` goes all the way down to the raw task.

One option I am thinking about is making periodic return a `Result`: the timer would not run forever in that case, but keep running until you return an error. This way we would be able to make it stop at som epoint

Reply:
Another option would be to require that the future passed in to the timer takes a reference to the timer itself so you have a chance to stop the timer once per interval.

Reply:
Yet another option that would make timers more general and useful would be to require that futures passed in return `Option<Duration/Instant>` so that the timer knows if the caller wants it to fire again. 

It would also make implementing jitter simpler.

Reply:
I don't like your first suggestion, it feels like a layering violation.

Taking an action based on the return value is a lot cleaner in my opinion.
If I go this route I may also change the name from `periodic` to `repeat`.

Another thing that is missing btw on this submission is a way to fire into a specific task queue which is different than the caller task queue, saving one indirection.